documentation/design.md

```diff
@@ -12,50 +10,53 @@ BuildPrediction is designed to work with the Graph API to make it possible to pr
-* MSBuild APIs used by BuildPrediction must be public; the MSBuild team makes a lot of effort to provide downlevel API compatibility, at least within a single major version number.
-* BuildPrediction initially targets .NET Framework for compatibility with Visual Studio and the Framework flavor of MSBuild. It may be updated to provide a .NET Standard flavor depending on the need of downstream consumers, e.g. use in a .Net Core build engine on Linux and Mac.
-* BuildPrediction must be shipped to users as a NuGet package, using standard SemVer versioning.
-* BuildPrediction implements a set of "standard" predictors maintained within the BuildPrediciton repo and shipped in the package, but may provide a pluggable model for externally written Predictors to be added.
-* BuildPrediction may provide a configuration mechanism to disable individual Predictors based on operational needs of a specific build. For example, a Predictor that is producing bad results in a specific repo might be completely turned off, or turned off in favor of a custom replacement Predictor that produces a better result for that repo.
-* BuildPrediction must be executed within the scope of a single repo encompassing a specific filesystem subdirectory called the RepositoryRoot and all subdirectories below that. When a build that includes Git submodules is in use, the RepositoryRoot is typically the subdirectory containing the outermost Git repo.
+* MSBuildPrediction is assumed to be downstream from the MSBuild Graph API and can take a dependency on part or all of the resulting graph object model.
+* MSBuildPrediction uses the Project as the unit of execution and caching. This matches the Graph API semantics, and matches the internal Microsoft build engine semantics. Note that this is in opposition to more complicated build models where Project files' Targets take cross-dependencies with the Targets of other Projects in a back-and-forth pattern, e.g. build "passes;" the Graph API model labels these as circular dependencies and requires that such Project graphs be unwound into a larger DAG.
+* MSBuildPrediction must produce a set of source code file input predictions for each Project in the graph. These predictions are intended to be - but are not required to be - as close as possible to the actual list of source files read, to ensure that an initial hash of the contents of these inputs produces a different hash-of-hashes to make lookups of cached outputs more exact. However, given the current nature of caching technology, underspecified source inputs produce more inefficient, but not broken, cache lookups, and overspecification is considered an anti-dependency when the predicted files are missing. The anti-dependency becomes a dependency on the file if in the next cache check the file appeared in the filesystem, and the hash of the newly appears file feeds into the cache lookup.
```

Comment:
>current nature of caching technology [](start = 402, length = 36)

linkify to: [https://github.com/microsoft/BuildXL/blob/master/Public/Src/Cache/README.md](https://github.com/microsoft/BuildXL/blob/master/Public/Src/Cache/README.md)